### PR TITLE
feat(profile): proxy public profile artifacts from s3

### PR DESCRIPTION
**Pull request by fayazg**
## Summary

Backend-only foundation for the public profile page (LFXV2-2631, part of the
myprofile → LFX Self-Serve cutover epic LFXV2-1948). Adds a server-side proxy
that reads the per-user public profile JSON artifact the upstream
`GenerateUserPublicProfile` flow publishes to S3, so the (later) public page can
render it SSR-safe without exposing S3 to the browser or dealing with CORS.

- **New public endpoint** `GET /public/api/profile/:username` — mounted under the
  `/public/api` stack (anonymous, `publicApiRateLimiter`, no auth). Confirmed out
  of scope of the impersonation service (mount order + identity-agnostic service).
- **`PublicProfileService`** — fetches `${bucket}/${username}.json`; maps 404/403 →
  not-found (null), non-OK/empty/invalid-JSON/timeout/network → `MicroserviceError`
  with appropriate status. Normalizes the artifact's public flag (`IsPublic` /
  `isPublic`) into a single `isPublic` boolean.
- **Controller** returns 404 for a missing profile, `{ isPublic: false }` (body
  withheld) for a private profile, and the full payload for a public one.
- **Username validation** — anchored pattern restricts the value before it's
  interpolated into the S3 object key (no path traversal / key probing).
- **Bucket URL is required config** — `PUBLIC_PROFILES_BUCKET_URL` env var, with no
  baked-in default. When unset/malformed/non-http, the app still boots and only this
  endpoint is inert (responds 503), so a misconfiguration surfaces loudly and in
  isolation rather than silently serving the wrong environment's data.
- Shared `PublicProfile` interface in `@lfx-one/shared`.
- 20 unit tests covering validation, config guards, status mapping, and flag
  normalization.

## Configuration

`PUBLIC_PROFILES_BUCKET_URL` must be set per environment to the base URL of the
bucket hosting the public profile artifacts. Handled entirely in code — no default
is baked in.

Part of LFXV2-2631

### CODE REVIEW

File: apps/lfx-one/src/server/services/public-profile.service.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        status: upstream.status,
+        status_text: upstream.statusText,
+        body,
+        username,
+      });
+      throw new MicroserviceError(`Public profile fetch failed: ${upstream.status} ${upstream.statusText}`, upstream.status, 'PUBLIC_PROFILE_FETCH_FAILED', {
+        operation,
+        service: PUBLIC_PROFILE_SERVICE_NAME,
+        errorBody: body,
+      });
```
[the remark this answers is not in the document]
Comment (fayazg) [reply apps/lfx-one/src/server/services/public-profile.service.ts]: Done — both sites now log only the upstream body length (`body_length`) instead of the raw body, and the raw `errorBody` payload has been removed from the `MicroserviceError` at the non-OK and invalid-JSON branches. Status/status-text are retained for diagnostics. Good catch — this is a public, unauthenticated endpoint whose artifact can carry PII, so keeping raw bodies out of logs is the right call.

See commit d4c86b1ba.

**Comment by fayazg** [comment]
## Review Feedback Addressed

Commit: d4c86b1ba

### Changes Made
- **apps/lfx-one/src/server/services/public-profile.service.ts**: at both the non-OK and invalid-JSON branches, log the upstream `body_length` instead of the raw body and remove the raw `errorBody` from the thrown `MicroserviceError`, so a public PII-bearing artifact can't leak into logs. Status/status-text retained for diagnostics. (per coderabbitai)

### Threads Resolved
1 of 1 unresolved threads addressed in this iteration.

File: apps/lfx-one/src/server/services/public-profile.service.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  return (process.env[PUBLIC_PROFILES_BUCKET_URL_ENV] || '').trim().replace(/\/+$/, '');
+}
+
+/**
+ * Normalizes the artifact's public flag to a boolean: an absent flag means public
+ * (published implies public); only an explicit `false` (IsPublic/isPublic) is private.
+ */
+export function resolvePublicFlag(parsed: Record<string, unknown>): boolean {
+  const raw = parsed['IsPublic'] ?? parsed['isPublic'];
+  return raw !== false;
```
[the remark this answers is not in the document]
Comment (fayazg) [reply apps/lfx-one/src/server/services/public-profile.service.ts]: Fixed in `bab216908`. `resolvePublicFlag` now fails closed — it releases the payload only when the flag is explicitly boolean `true`:

```ts
const raw = parsed['IsPublic'] ?? parsed['isPublic'];
return raw === true;
```

A missing or non-boolean flag (`"false"`, `0`, `{}`) now resolves to `false`, so the controller returns `{ isPublic: false }` and withholds the payload. Added unit coverage for the absent / non-boolean / stray-camelCase-`true` cases, plus a new controller test asserting the private branch never leaks PII.

File: apps/lfx-one/src/server/services/public-profile.service.ts [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    try {
+      parsed = JSON.parse(rawBody) as Record<string, unknown>;
+    } catch (error: unknown) {
+      const truncatedBody = rawBody.slice(0, UPSTREAM_ERROR_BODY_LIMIT);
+      logger.warning(req, operation, 'Public profile artifact was invalid JSON', { err: error, body: truncatedBody, username });
+      throw new MicroserviceError('Public profile fetch failed: invalid JSON response from upstream', 502, 'UPSTREAM_INVALID_RESPONSE', {
+        operation,
+        service: PUBLIC_PROFILE_SERVICE_NAME,
+        errorBody: truncatedBody,
+      });
```
[the remark this answers is not in the document]
Comment (fayazg) [reply apps/lfx-one/src/server/services/public-profile.service.ts:161]: Already addressed in `d4c86b1` (landed before this thread was posted, which is why it's marked outdated). The service now logs only `body_length` — never the raw body — on both the non-OK and invalid-JSON paths, and the `MicroserviceError` for those paths carries no `errorBody`, so the centralized handler has no artifact content to log either. No raw profile data reaches the logs.

File: apps/lfx-one/src/server/controllers/public-profile.controller.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+          operation: 'get_public_profile',
+          service: 'public_profile_controller',
+          path: req.path,
+        });
+      }
+
+      logger.success(req, 'get_public_profile', startTime, { username, is_public: profile.isPublic });
+
+      if (!profile.isPublic) {
+        res.json({ isPublic: false });
```
[the remark this answers is not in the document]
Comment (fayazg) [reply apps/lfx-one/src/server/controllers/public-profile.controller.ts:36]: Done in `bab216908`. Added `public-profile.controller.spec.ts` following the `public-meeting.controller.spec.ts` pattern — it stubs `PublicProfileService` and asserts:

- **private profile** → responds with exactly `{ isPublic: false }`, and asserts none of the withheld payload fields (name, bio, etc.) appear anywhere in the response
- **public profile** → responds with the full payload
- **missing profile** (service returns `null`) → forwards a 404 `ResourceNotFoundError`
- **service error** → forwards to `next(error)`

Good catch — the controller is the actual exposure boundary, so it now has direct coverage.

File: apps/lfx-one/src/server/services/public-profile.service.ts [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    } catch (error: unknown) {
+      // Log only the body length, never the raw body — it may hold private profile data.
+      logger.warning(req, operation, 'Public profile artifact was invalid JSON', { err: error, body_length: rawBody.length, username });
+      throw new MicroserviceError('Public profile fetch failed: invalid JSON response from upstream', 502, 'UPSTREAM_INVALID_RESPONSE', {
+        operation,
+        service: PUBLIC_PROFILE_SERVICE_NAME,
+      });
+    }
+
+    return { ...parsed, isPublic: resolvePublicFlag(parsed) } as PublicProfile;
```
[the remark this answers is not in the document]
Comment (fayazg) [reply apps/lfx-one/src/server/services/public-profile.service.ts]: Fixed in `bab216908`. After a successful parse the service now guards the shape before touching the flag:

```ts
if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed)) {
  logger.warning(req, operation, 'Public profile artifact was not a JSON object', { body_length: rawBody.length, username });
  throw new MicroserviceError('Public profile fetch failed: unexpected response shape from upstream', 502, 'UPSTREAM_INVALID_RESPONSE', {
    operation,
    service: PUBLIC_PROFILE_SERVICE_NAME,
  });
}
```

So `null`, arrays, and primitives now map to a 502 `MicroserviceError` (consistent with the empty / invalid-JSON handling) instead of throwing on the flag lookup or being spread into a "public" payload. Added tests for the null / array / primitive cases.

**Comment by fayazg** [comment]
## Review Feedback Addressed

Commit: `bab216908`

### Changes Made
- **`public-profile.service.ts`** — fail closed on the privacy gate: `resolvePublicFlag` releases the payload only for an explicit boolean `true`; a missing or non-boolean flag now resolves to `false` (per Copilot)
- **`public-profile.service.ts`** — reject non-object JSON (`null` / array / primitive) with a 502 `MicroserviceError` instead of crashing on the flag lookup or spreading a corrupt payload into a "public" response (per Cursor Bugbot)
- **`public-profile.controller.spec.ts`** _(new)_ — controller tests for the data-exposure boundary: private → exactly `{ isPublic: false }` (asserts none of the withheld fields leak), public → full payload, missing → 404, error → `next(error)` (per Copilot)
- **`public-profile.interface.ts`** — documented the fail-closed `isPublic` semantics
- **`public-profile.service.spec.ts`** — added fail-closed flag handling + non-object artifact body coverage

### No Change Needed
- **Raw artifact body in logs** — already addressed in `d4c86b1` (thread was outdated): only `body_length` is logged on the non-OK / invalid-JSON paths, and the `MicroserviceError` carries no `errorBody`, so no profile content reaches the logs (flagged by Copilot)

### Threads Resolved
4 of 4 unresolved threads addressed and resolved.

File: apps/lfx-one/src/server/services/public-profile.service.ts [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        service: PUBLIC_PROFILE_SERVICE_NAME,
+      });
+    }
+
+    let parsed: unknown;
+    try {
+      parsed = JSON.parse(rawBody);
+    } catch (error: unknown) {
+      // Log only the body length, never the raw body — it may hold private profile data.
+      logger.warning(req, operation, 'Public profile artifact was invalid JSON', { err: error, body_length: rawBody.length, username });
```
[the remark this answers is not in the document]
Comment (fayazg) [reply apps/lfx-one/src/server/services/public-profile.service.ts]: Fixed in 64f5193d1 — the invalid-JSON branch now logs only `error_name` (e.g. `SyntaxError`) and `body_length`, and no longer passes the error object to the logger, so the `SyntaxError`'s embedded input snippet can't reach the logs. Applied the same treatment to the pre-response network-error log for consistency.

File: apps/lfx-one/src/server/services/public-profile.service.ts [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+/**
+ * Normalizes the artifact's public flag to a boolean. This is the sole privacy gate for an
+ * anonymous endpoint, so it fails closed: the payload is released only when the upstream flag
+ * (IsPublic/isPublic) is explicitly boolean `true`. A missing or non-boolean flag resolves to
+ * `false` — artifacts exist for private profiles too, so a dropped/malformed flag must not leak.
+ */
+export function resolvePublicFlag(parsed: Record<string, unknown>): boolean {
+  const raw = parsed['IsPublic'] ?? parsed['isPublic'];
+  return raw === true;
```
[the remark this answers is not in the document]
Comment (fayazg) [reply apps/lfx-one/src/server/services/public-profile.service.ts]: Fixed in 64f5193d1 — `resolvePublicFlag` is now fully fail-closed: it releases the payload only when at least one flag casing is present **and** every present casing is exactly boolean `true`. So `{ IsPublic: true, isPublic: false }` and `{ IsPublic: null, isPublic: true }` both resolve to private now. Added reciprocal-conflict and null-casing test coverage.

As related hardening on this anonymous endpoint, raw network diagnostics (ENOTFOUND/TLS) are no longer forwarded to callers — the service now throws a stable `UPSTREAM_UNREACHABLE`/502 and keeps the detail in the server log.

**Comment by fayazg** [comment]
## Review Feedback Addressed

Commit: 64f5193d1

### Changes Made
- **public-profile.service.ts**: invalid-JSON parse now logs only `error_name` + `body_length`, not the error object — Node's `SyntaxError` message embeds an input snippet that could carry private-profile PII (per cursor[bot])
- **public-profile.service.ts**: `resolvePublicFlag` is now fully fail-closed — releases the payload only when at least one flag casing is present and every present casing is exactly boolean `true`, so a conflicting (`{ IsPublic: true, isPublic: false }`) or null-casing (`{ IsPublic: null, isPublic: true }`) artifact resolves to private (per copilot[bot])
- **public-profile.service.ts**: network-failure path now throws a stable `Public profile fetch failed: upstream unreachable` / `UPSTREAM_UNREACHABLE` / 502 instead of forwarding raw fetch diagnostics (ENOTFOUND, ECONNREFUSED, TLS) to anonymous callers; the diagnostics stay in the server log (per copilot[bot])
- **public-profile.service.ts**: the bucket **base** URL is now validated (rather than the combined key URL), so a `https://` misconfig collapses to `https:` and fails 503 instead of treating the username as the host and fetching (per copilot[bot])
- **public-profile.controller.spec.ts**: removed the module-level `USERNAME` constant (moved to a `buildReqRes` default param) and typed the Express fixtures, dropping both `as any` casts (per coderabbitai[bot])
- **public-profile.service.spec.ts**: added fail-closed conflict/null-casing coverage and a no-host bucket (`https://`) case

### No Change Needed (already addressed in earlier commits)
- **public-profile.controller.ts private-path test**: `public-profile.controller.spec.ts` already asserts the response is exactly `{ isPublic: false }` with no withheld field leaking (flagged by copilot[bot])
- **JSON.parse null/array/primitive shape**: already guarded in `public-profile.service.ts` (rejects non-object bodies as `UPSTREAM_INVALID_RESPONSE`) with tests (flagged by copilot[bot])

### Threads Resolved
2 of 2 unresolved threads addressed in this iteration.

**Review by MRashad26 — APPROVED** [review]
Well-designed public-endpoint proxy with strong privacy and security fundamentals. Username pattern + `encodeURIComponent` close path-traversal before S3 key construction; `resolvePublicFlag` fails closed for any missing/non-boolean/conflicting flag; the controller is the exposure boundary and never forwards withheld fields (test verifies with serialized inspection). Bucket URL validated as a standalone base URL, not combined with the username — correctly catches the empty-authority edge case. No baked-in default means misconfiguration is loud and isolated. All code-enforcer checks pass.\n\n**Nit:** `private publicProfileService` in `PublicProfileController` should be `private readonly` to match the convention used in all other controllers.

**Comment by dealako** [comment]
Hey @fayazg — nice work here. This is a clean, well-scoped backend addition: a single anonymous S3 proxy endpoint with fail-closed privacy handling, anchored username validation ahead of the S3 key, and a config guard that degrades only this one route instead of the whole app. The iterative review history (raw-body logging scrubbed, fail-closed flag semantics, sanitized error logs) shows the design tightened up nicely across rounds, and the 20 unit tests cover the validation/status-mapping/flag-normalization surface well.

Ran a 3-way split review (security & data privacy / correctness, performance & test coverage / style, API compliance & docs).

### 🔴 Blocking: 0

### 🟡 Minor: 0

### ⚪ Nit: 2
- **No Cache-Control on the public profile response** (`public-profile.controller.ts`) — every request re-fetches from S3 with no cache headers, so a popular profile re-hits S3 on every load. Worth a `Cache-Control: public, max-age=...` at some point, but it's not a regression — the sibling `public-meeting.controller.ts` follows the same no-header pattern, so this is pre-existing convention rather than something this PR introduced.
- **Backend architecture README not updated** (`docs/architecture/backend/README.md:155,173`) — still lists "public meetings and public committees" as the only public surfaces; doesn't mention the new `public-profile` domain now under `/public/api/*`.

### ❔ Question: 0

**Security & data privacy:** no findings — traced username → validation → S3 key construction → status mapping → response, and log/error sinks. The already-addressed CodeRabbit finding (raw upstream body in error logs) and Cursor Bugbot flag both check out as resolved in the current code (`error_name` + `body_length` only). Fail-closed behavior on `resolvePublicFlag` and the private-profile response stripping (`{ isPublic: false }` only) both hold up.

**Reconciliation with bots:** CodeRabbit's one actionable comment (raw error-body logging) and Cursor Bugbot's flagged issue were both already resolved by the author's follow-up commits (`d4c86b1`, `bab2169`, `64f5193`) before this review — confirmed against current code, not just the PR description.

### ✅ Approved

Given zero blocking/security findings and only two nits (neither a regression), this is good to merge as-is; the nits are optional follow-ups.

**Review by dealako — APPROVED** [review]
Approved — see the consolidated review comment above for full findings (0 blocking, 0 security findings, 2 non-regression nits).
